Migrate to Apple-native strings catalog and codegen LocalizedStringResources with xcstrings-tool-plugin

tyiu opened the pull request:
Closes: https://github.com/planetary-social/nos/issues/718

**Previously**,
English strings were written in `Nos/Assets/Localization/Localized.swift` on the `Localized` enum, and then `Generated.strings` were generated for each locale that Nos supports, which could be localized and read in code.

**Now**,
`Nos/Assets/Localization/*.xcstrings` (string catalog) files are used instead. String catalogs were introduced in Xcode 15 and they are supported natively, including variable substitution and localized plural rules. Each xcstrings file contains the translations of all locales. Crowdin has support for xcstrings files, and will write translations into the same file. I opted to use an MIT-licensed open source plugin called [xcstrings-tool](https://github.com/liamnichols/xcstrings-tool) that transforms xcstrings files to generated code, so that we can reference `LocalizedStringResources` in a type-safe way.

I copied all the current translations from Crowdin to the string catalog files. There was a bit of manual work here so hopefully I got it all right.

Someone who has access to the Crowdin project will need to make sure the integration works fine with these changes. I did some testing on my local Crowdin project and it seemed to work. Keep in mind that the history of the past translation work and translator attribution will probably get wiped out.

There seems to be an issue where Xcode formats the JSON in the *.xcstrings files with a space after the JSON key, but Crowdin does not. There will likely be some noise flip-flopping back and forth between the spacing discrepancies until either one of them decides to change how they write xcstrings files to be consistent with the other.

| English | German | Chinese (Traditional) |
| ----- | ---------- | -------- |
| ![Simulator Screenshot - iPhone 15 Pro - 2023-12-14 at 10 33 57 Medium](https://github.com/planetary-social/nos/assets/963907/439454d1-9ca0-49f3-a6e7-bc980452aa01) | ![Simulator Screenshot - iPhone 15 Pro - 2023-12-14 at 10 33 40 Medium](https://github.com/planetary-social/nos/assets/963907/bf7b21f5-1ee6-4a9e-8c48-b7abc3e6fcbe) | ![Simulator Screenshot - iPhone 15 Pro - 2023-12-14 at 10 42 46 Medium](https://github.com/planetary-social/nos/assets/963907/f8adb7aa-8d08-4051-b16c-56ee123ed0f0) |

tyiu commented on Nos/Views/AuthorStoryView.swift:
```diff
@@ -251,9 +251,7 @@ fileprivate struct BottomOverlay: View {
             return nil
         }
-        let replyCount = replyCount
-        let localized = replyCount == 1 ? Localized.Reply.one : Localized.Reply.many
```
This logic did not work for locales that support different pluralization rules than just `one` and `many`.

tyiu commented on Nos/Views/RepostButton.swift:
```diff
@@ -57,16 +57,16 @@ struct RepostButton: View {
             .padding(.vertical, 12)
         }
         .disabled(tapped)
-        .confirmationDialog("Repost", isPresented: $shouldConfirmRepost) { 
```
Hardcoded string that was not localized.

tyiu commented on Nos/Views/RepostButton.swift:
```diff
@@ -57,16 +57,16 @@ struct RepostButton: View {
         }
         .disabled(tapped)
-        .confirmationDialog("Repost", isPresented: $shouldConfirmRepost) { 
-            Button("Repost") { 
```
Hardcoded string that was not localized.

tyiu commented on Nos/Views/RepostButton.swift:
```diff
@@ -57,16 +57,16 @@ struct RepostButton: View {
                 tapped = false
             }
         }
-        .confirmationDialog("Delete repost", isPresented: $shouldConfirmDelete) { 
```
Hardcoded string that was not localized.

tyiu commented on Nos/Views/RepostButton.swift:
```diff
@@ -57,16 +57,16 @@ struct RepostButton: View {
             }
         }
-        .confirmationDialog("Delete repost", isPresented: $shouldConfirmDelete) { 
-            Button("Delete repost", role: .destructive) { 
```
Hardcoded string that was not localized.

tyiu commented on Nos/Views/NoteCard.swift:
```diff
@@ -42,9 +42,7 @@ struct NoteCard: View {
             return nil
         }
-        let replyCount = replyCount
-        let localized = replyCount == 1 ? Localized.Reply.one : Localized.Reply.many
```
This logic did not work for locales that support different pluralization rules than just `one` and `many`.

tyiu commented on Nos/Views/ProfileKnownFollowersView.swift:
```diff
@@ -17,28 +17,16 @@ struct ProfileKnownFollowersView: View {
-                        "two": second.safeName,
-                        "count": "\(followers.count - 2)"
-                    ]))
+                    Text(.localizable.followedByTwoAndMore(first.safeName, second.safeName, followers.count - 2))
```
I wasn't able to figure out if named arguments could be supported. Leaving that as an open question.

tyiu commented on crowdin.yml:
```diff
@@ -1,3 +1,13 @@
-    translation: /Nos/Assets/Localization/%osx_code%/%original_file_name%
+  - source: /Nos/Assets/Localization/ImagePicker.xcstrings
+    translation: /Nos/Assets/Localization/ImagePicker.xcstrings
+    multilingual: 1
```
Indicates that the file contains translations for multiple locales.

tyiu commented:
Rebased on the `main` branch and resolved the conflicts. I also brought in new Persian translations that came in after I created this pull request.

mplorentz reviewed (commented):
Wow, this is a huge contribution and a big improvement over the previous system! Regarding the history here: I actually don't know why we were using a handmade `ExportStrings.sh` to do localization. It was set up for the Planetary app before I started, and when we started Nos we copied a lot of the strings over and with it came this system. 

@martindsq can you give this a proper review since you manage the Crowdin integration and have a lot more expertise than me in this area?

tyiu commented:
>There seems to be an issue where Xcode formats the JSON in the *.xcstrings files with a space after the JSON key, but Crowdin does not. There will likely be some noise flip-flopping back and forth between the spacing discrepancies until either one of them decides to change how they write xcstrings files to be consistent with the other. This may cause merge conflicts, not sure what the best path forward is. I've filed an issue here and they've said that they're working on a fix: https://community.crowdin.com/t/apple-xcstrings-string-catalog-format-discrepancy-between-crowdin-and-xcode/5937

Crowdin has fixed the issue. I'm removing this blurb from the PR description.

tyiu commented:
- [ ] This PR should be rebased and adjusted after https://github.com/planetary-social/nos/pull/742 is merged to `main` with its localization code and string changes.

mplorentz commented:
> * [ ]  This PR should be rebased and adjusted after [Tweak warning message to support localization #742](https://github.com/planetary-social/nos/pull/742) is merged to `main` with its localization code and string changes.

Yeah sorry about that. Just clicked merge on that PR. Also we have some tests failing on `main`, I will be addressing them tomorrow.

tyiu commented:
> > * [ ]   This PR should be rebased and adjusted after [Tweak warning message to support localization #742](https://github.com/planetary-social/nos/pull/742) is merged to `main` with its localization code and string changes.
> 
> Yeah sorry about that. Just clicked merge on that PR. Also we have some tests failing on `main`, I will be addressing them tomorrow.

No worries. I just rebased and resolved the conflicts..

tyiu commented on Nos/Views/PreviewData.swift:
```diff
@@ -55,7 +55,7 @@ struct PreviewData {
     
     lazy var bob: Author = {
-        let author = Author(context: previewContext)
+        let author = try! Author.findOrCreate(by: KeyFixture.bob.publicKeyHex, context: previewContext)
```
I think this was a bug that `bob` wasn't in Core Data. `alice` and `eve` were though, so I copied what they did.

tyiu commented on Nos/Views/WarningView.swift:
```diff
@@ -157,13 +157,13 @@ struct ContentWarningMessage: View {
     // Assuming there's a property or method 'safeName' in 'Author' that safely returns the author's name.
     private var firstAuthorSafeName: String {
         // Getting the safe name of the first author. Adjust according to your actual data structure.
-        reports.first?.author?.safeName ?? "Unknown Author"
```
Pulled out `Unknown Author` to localize it.

tyiu commented on Nos/Views/WarningView.swift:
```diff
@@ -157,13 +157,13 @@ struct ContentWarningMessage: View {
     
     private var reason: String {
         // Extract content from reports and remove empty content
-        let contents = reports.compactMap { (($0.content?.isEmpty) != nil) ? nil : $0.content }
```
This was a bug. It would always return `nil`.

tyiu commented:
Rebased on the `main` branch, resolved conflicts, and added new translations for Chinese Simplified, Chinese Traditional, and German locales.
